### DISCUSSION TITLE
Weighted Bézier curve to arbitrary height

### DISCUSSION THREAD

**Discussion by Wilker-uwu** [NicolasKlenert/enterpolation#29] [Show and tell]
behold! my very own piece of hardcoded spaghetti i've made using this library!

```rust
    //[...lots of unrelated code upwards...]
    
```rust
    #[func]
    pub fn get_throttle_input(&self) -> f32 {
        Input::singleton().get_axis("yaw_left".into(), "yaw_right".into())
    }

    #[func]
    pub fn get_target_speed(&mut self) -> f32 {
        let t = remap(self.get_throttle_input(),-1.0,1.0,0.0,1.0);
        let w = 1.0;

        let bez = Bezier::builder().elements_with_weights([
                (self.minimum, w),
                (self.stable, w),
                (self.stable, w),
                (self.stable, w),
                (self.maximum, w)])
            .domain(0.0, 1.0)
            .dynamic()
            .build()
            .unwrap();
        let samples = (0..7).map(|i| bez.gen(i as f32 / 7.0)).collect::<Vec<f32>>();
        godot_print!("curve: {:?},", samples);
        let i = bez.gen(t);
        godot_print!("i: {},", i);
        i
    }
```

after 2 days and a lot of pain, this library has saved me :heart:  
now excuse me while i look away and never touch this section of the code ever again
###### update: please please PLEEEASE let me hope i never have to touch this piece of shenanigan code again. HOWEVER!! if you do know any improvements to be made, let me know!

**Comment by NicolasKlenert** [comment]
So for everybody who still want to touch this section of code, some things to notice:

- if all of your weights have the same value, you don't need weights at all
- assuming that samples should go from the first value to the last value, the range `(0..=7)` or `(0..8)` should be used.
- (then) more idomtic code: `let samples = bez.take(8).collect::<Vec<f32>>();;`
- performance improvement: instead of `.domain(0.0, 1.0)`, one can use `.normalized()`. It also makes the resulting bezier type simpler.
- in the above case however, a domain of `[-1.0,1.0]` is wished for, so the remap function call can be deleted and instead `.domain(-1.0,1.0)` can be used.
- performance improvement: instead of `dynamic()`, `constant()` can be used (because the elements are inserted with an array, which size in known at compile time).
- performance improvement: using `new_unchecked()` to skip all error checks, as all curves possible by the code are valid curves (and no error handling exist anyways). This however creates code which is not as clean (and needs some type information), so there is some drawback.


All in all we have either (assuming you want different weights for the different elements):
```rust
#[func]
    pub fn get_target_speed(&mut self) -> f32 {
        let w = [1.0,1.0,2.0,1.0,1.0];

        let bez = Bezier::builder().elements_with_weights([
                (self.minimum, w[0]),
                (self.stable, w[1]),
                (self.stable, w[2]),
                (self.stable, w[3]),
                (self.maximum, w[4])])
            .domain(-1.0, 1.0)
            .constant()
            .build()
            .unwrap();
        let samples = bez.take(8).collect::<Vec<f32>>();
        godot_print!("curve: {:?},", samples);
        let i = bez.gen(self.get_throttle_input());
        godot_print!("i: {},", i);
        i
    }
```

or the optimised, but a bit uglier, version:
```rust
#[func]
    pub fn get_target_speed(&mut self) -> f32 {
        let w = [1.0,1.0,2.0,1.0,1.0];

       let weighted_bez =Bezier::<f32, Weights<_>, ConstSpace<Homogeneous<_, _>, 5>>::new_unchecked(
            Weights::new([
                (self.minimum, w[0]),
                (self.stable, w[1]),
                (self.stable, w[2]),
                (self.stable, w[3]),
                (self.maximum, w[4])
            ]),
            ConstSpace::new(),
        );
        // transform the weight pairs of the curve into a real number, so it behaves like a normal curve:
        let normalized_bez = Weighted::new(weighted_bez);
        // to change from the normalized domain to another domain:
        let bez = TransformInput::normalized_to_domain(normalized_bez, -1.0, 1.0);

        let samples = bez.take(8).collect::<Vec<f32>>();
        godot_print!("curve: {:?},", samples);
        let i = bez.gen(self.get_throttle_input());
        godot_print!("i: {},", i);
        i
    }
```

**Reply by Wilker-uwu** [reply]
thank you so much! :heart: definitely yoinking those :p

**Reply by Wilker-uwu** [reply]
one question: why do the approaches requires me to subtract the result of `get_throttle_input()` by 1 before feeding it into the `gen(t)` function? it's cool that it's still easy to use your solution as-is but i'm still confused

again thank you for the help <3

**Reply by NicolasKlenert** [reply]
Depends on if I understood your code correctly. I thought you used a remap function because the values of `throttled_input()` returns a value between -1.0 and 1.0. The domain of the curve basically describes what the curve expects as input.

Does this answer your question?

**Reply by Wilker-uwu** [reply]
yeah that's pretty much it. thank you ^^ might be something on my side when placing the inputs, but that i'll be checking later ^^' for now it works :3
